### PR TITLE
Fix L2 normalization fusion for multi-axis reductions

### LINKED ISSUE #118181: TFLite produces incorrect output for tf.math.l2_normalize with default axis on shape [N, 1]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.21.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

3.11

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

TensorFlow Lite produces a different result from Keras for tf.math.l2_normalize on a tensor with shape [3, 1].

The Keras model applies tf.math.l2_normalize(v) to the input. Keras matches the NumPy reference using the full tensor norm, but the converted TFLite model returns [1.0, 1.0, 1.0].

In the reproducer below, the expected output is approximately [0.3801878, 0.5974380, 0.7060631], while TFLite returns all ones.

Expected behavior?

The converted TFLite model should match the Keras output for tf.math.l2_normalize with the same input shape and default axis behavior.

### Standalone code to reproduce the issue

```shell
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "-1"

import numpy as np
import tensorflow as tf


class M(tf.keras.Model):
    def call(self, v):
        return tf.math.l2_normalize(v)


X = np.array([[7.0], [11.0], [13.0]], dtype=np.float32)

model = M()
keras_out = model(tf.constant(X)).numpy()
ref = X / np.linalg.norm(X)

converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()

interpreter = tf.lite.Interpreter(model_content=tflite_model)
input_idx = interpreter.get_input_details()[0]["index"]
interpreter.resize_tensor_input(input_idx, list(X.shape))
interpreter.allocate_tensors()
interpreter.set_tensor(input_idx, X)
interpreter.invoke()

tflite_out = interpreter.get_tensor(interpreter.get_output_details()[0]["index"])

print(f"ref={ref.ravel().tolist()}")
print(f"keras={keras_out.ravel().tolist()}")
print(f"tflite={tflite_out.ravel().tolist()}")

if np.allclose(keras_out, ref) and not np.allclose(tflite_out, keras_out):
    print("BUG REPRODUCED: TFLite l2_normalize output differs from Keras/reference")
else:
    print("not reproduced")
```

### Relevant log output

```shell
ref=[0.3801878094673157, 0.5974379777908325, 0.7060630917549133]
keras=[0.3801878094673157, 0.5974379777908325, 0.7060630321502686]
tflite=[1.0, 1.0, 1.0]
BUG REPRODUCED: TFLite l2_normalize output differs from Keras/reference
```

### PR DESCRIPTION
**Pull request by laolvfan**
## Summary

  This change prevents the L2 normalization rewrite from fusing reductions over
  multiple axes.

  `TFL_L2_NORMALIZATION` normalizes only along the trailing dimension. The existing
  constraint also accepted an axis list containing every input dimension, which could
  change a global L2 normalization into a per-trailing-dimension normalization.

  The updated constraint:

  - requires exactly one reduction axis;
  - accepts `rank - 1` or its equivalent negative index `-1`;
  - leaves multi-axis normalization patterns unfused.

  A regression test verifies that the `square -> sum -> rsqrt -> mul` sequence is
  preserved for `axis = [0, 1]`.

  ## Testing

  ```text
  bazel test //tensorflow/compiler/mlir/lite/tests:optimize.mlir.test
```

  Fixes #118181

  Related to google-ai-edge/LiteRT#9212.
  This supersedes my earlier attempt in google-ai-edge/LiteRT#9464.

### CODE REVIEW

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thank you for the PR!

Our automated test suite reported a C++ formatting issue. Could you please format the code (e.g., using `clang-format` or applying the patch below)?

```patch
diff --git a/tensorflow/compiler/mlir/lite/transforms/optimize_pass.cc b/tensorflow/compiler/mlir/lite/transforms/optimize_pass.cc
index d906e3f7055..cd655a64d1f 100644
--- a/tensorflow/compiler/mlir/lite/transforms/optimize_pass.cc
+++ b/tensorflow/compiler/mlir/lite/transforms/optimize_pass.cc
@@ -124,8 +124,7 @@ bool L2NormalizeReduceAxis(Value sq_op, DenseElementsAttr axis) {
   }
 
   const int reduce_axis = *axis.getValues<int>().begin();
-  const int64_t rank =
-      mlir::cast<ShapedType>(sq_op.getType()).getRank();
+  const int64_t rank = mlir::cast<ShapedType>(sq_op.getType()).getRank();
   return reduce_axis == rank - 1 || reduce_axis == -1;
 }
```

**Comment by laolvfan** [comment]
Hi @dmiltr3, thanks for your patience. I applied the suggested formatting change.

**Review by dmiltr3 — APPROVED** [review]
Thank you for the contribution!

Internal presubmits have verified the changes and everything looks good. Approving this PR.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thank you for your contribution!

While testing your change against the internal automated test suite, we encountered regressions in quantized models (such as Gazenet):
- `Node number 21 (RSQRT) failed to prepare: Type UINT8 is unsupported by op Rsqrt.`
- `INTERNAL: Failed allocating tensors.`

### Root Cause
`TFL_L2_NORMALIZATION` normalizes along the trailing dimension. However, when an input tensor has static leading dimensions equal to 1 (e.g. `[1, 3]` or `[1, 1, 1, 3]`), reducing across all dimensions is mathematically identical to normalizing along the trailing dimension because the outer product size is 1.

By strictly requiring `axis.getNumElements() == 1`, models reducing across all axes with unit leading dimensions are no longer fused into `tfl.l2_normalization`. The unfused `square -> sum -> rsqrt -> mul` sequence undergoes quantization where `rsqrt` receives a `UINT8` tensor, which is not supported by TFLite's `Rsqrt` kernel at runtime.

### Suggested Fix
We can continue allowing reductions across all dimensions when all leading dimensions are statically equal to 1, while disallowing multi-axis reductions when leading dimensions are greater than 1 (such as your `tensor<2x3xf32>` test case):

```patch
diff --git a/tensorflow/compiler/mlir/lite/transforms/optimize_pass.cc b/tensorflow/compiler/mlir/lite/transforms/optimize_pass.cc
--- a/tensorflow/compiler/mlir/lite/transforms/optimize_pass.cc
+++ b/tensorflow/compiler/mlir/lite/transforms/optimize_pass.cc
@@ -123,21 +123,32 @@ bool L2NormalizeReduceAxis(Value sq_op, DenseElementsAttr axis) {
   if (axis.getNumElements() == 0) {
     return false;
   }
-  if (mlir::cast<ShapedType>(sq_op.getType()).getRank() - 1 ==
-          *axis.getValues<int>().begin() ||
-      *axis.getValues<int>().begin() == -1) {
-    return true;
+  auto shape = mlir::dyn_cast<ShapedType>(sq_op.getType());
+  if (!shape || !shape.hasRank()) {
+    return false;
   }
-  if (mlir::cast<ShapedType>(sq_op.getType()).getRank() !=
-      axis.getNumElements()) {
+  const int64_t rank = shape.getRank();
+  if (axis.getNumElements() == 1) {
+    const int reduce_axis = *axis.getValues<int>().begin();
+    return reduce_axis == rank - 1 || reduce_axis == -1;
+  }
+  if (shape.getNumElements() == 0 || !shape.hasStaticShape()) {
     return false;
   }
-  auto shape = mlir::cast<ShapedType>(sq_op.getType());
+  if (axis.getNumElements() != rank) {
+    return false;
+  }
   SmallVector<int, 4> elems{axis.getValues<int>().begin(),
                             axis.getValues<int>().end()};
-  for (int i = 0; i < shape.getRank(); ++i) {
+  for (int i = 0; i < rank; ++i) {
     if (i != elems[i]) return false;
   }
+  // Reducing across all dimensions is only equivalent to TFL L2 normalization
+  // (which reduces along the trailing dimension) if all leading dimensions
+  // have size 1.
+  for (int i = 0; i < rank - 1; ++i) {
+    if (shape.getDimSize(i) != 1) return false;
+  }
   return true;
 }
```

Could you please update the PR with this adjustment? Thank you!

**Comment by laolvfan** [comment]
Hi @dmiltr3, thanks for the detailed feedback.

  I updated the implementation to preserve all-axis L2 normalization fusion when all leading dimensions are statically equal to 1, while continuing to reject unsafe multi-axis reductions such as tensor<2x3xf32>.

  I also added a regression test for the safe tensor<1x3xf32> case. The following test passes locally:  ```//tensorflow/compiler/mlir/lite/tests:optimize.mlir.test```

The Windows CI failure appears unrelated to this change. Could you please take another look? Thanks!

**Review by dmiltr3 — APPROVED** [review]
Thank you for updating the PR with the safe tensor unit leading dimensions enhancement and the regression test!

The internal presubmit tests have verified the changes and everything looks good. Approving this PR.
